drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,524 @@
+
+		if (mode == GPIO_INT_MODE_DISABLED) {
+			irq->falling &= ~BIT(pin);
+			irq->rising &= ~BIT(pin);
+		} else { /* GPIO_INT_MODE_EDGE */
+			if (trig == GPIO_INT_TRIG_BOTH) {
+				irq->falling |= BIT(pin);
+				irq->rising |= BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_LOW) {
+				irq->falling |= BIT(pin);
+				irq->rising &= ~BIT(pin);
+			} else if (trig == GPIO_INT_TRIG_HIGH) {
+				irq->falling &= ~BIT(pin);
+				irq->rising |= BIT(pin);
+			}
+		}
+		k_sem_give(&drv_data->lock);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int gpio_tca9538_manage_callback(const struct device *dev,
+					struct gpio_callback *callback,
+					bool set)
+{
+	struct tca9538_drv_data *data = dev->data;
+
+	return gpio_manage_callback(&data->cb, callback, set);
+}
+
+/**
+ * @brief Initialization function of TCA9538
+ *
+ * This sets initial input/ output configuration and output states.
+ * The interrupt is configured if this is enabled.
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_tca9538_init(const struct device *dev)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	int rc = 0;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if (!device_is_ready(cfg->i2c_dev)) {
+		LOG_ERR("I2C device not found");
+		goto out;
+	}
+
+	/* Do an initial read, this clears the interrupt pin and sets
+	 * up the initial value of the pin state input data.
+	 */
+	rc = update_input(dev);
+	if (rc) {
+		goto out;
+	}
+
+	if (cfg->interrupt_enabled) {
+		if (!device_is_ready(cfg->gpio_int.port)) {
+			LOG_ERR("Cannot get pointer to gpio interrupt device");
+			rc = -EINVAL;
+			goto out;
+		}
+
+		drv_data->dev = dev;
+
+		k_work_init(&drv_data->work, gpio_tca9538_work_handler);
+
+		rc = gpio_pin_configure_dt(&cfg->gpio_int, GPIO_INPUT);
+		if (rc) {
+			goto out;
+		}
+
+		rc = gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
+						GPIO_INT_EDGE_TO_ACTIVE);
+		if (rc) {
+			goto out;
+		}
+
+		gpio_init_callback(&drv_data->gpio_cb,
+					gpio_tca9538_init_cb,
+					BIT(cfg->gpio_int.pin));
+
+		rc = gpio_add_callback(cfg->gpio_int.port,
+					&drv_data->gpio_cb);
+	}
+out:
+	if (rc) {
+		LOG_ERR("%s init failed: %d", dev->name, rc);
+	} else {
+		LOG_INF("%s init ok", dev->name);
+	}
+	k_sem_give(&drv_data->lock);
+	return rc;
+}
+
+static const struct gpio_driver_api api_table = {
+	.pin_configure = gpio_tca9538_config,
+	.port_get_raw = gpio_tca9538_port_read,
+	.port_set_masked_raw = gpio_tca9538_port_set_masked,
+	.port_set_bits_raw = gpio_tca9538_port_set_bits,
+	.port_clear_bits_raw = gpio_tca9538_port_clear_bits,
+	.port_toggle_bits = gpio_tca9538_port_toggle_bits,
+	.pin_interrupt_configure = gpio_tca9538_pin_interrupt_configure,
+	.manage_callback = gpio_tca9538_manage_callback,
+};
+
+#define INST_DT_TCA9538(inst, t) DT_INST(inst, ti_tca9538##t)
+
+/* Refer to https://www.ti.com/lit/gpn/tca9538
+ * for default register setup.
+ */
```

Comment:
The URLs tend to get outdated fairly quickly, let's maybe remove it.

Reply:
Removed.